File: python/src/log.py
from logging import DEBUG, INFO, WARNING, ERROR
from sys import stderr
from typing import Any

from colorama import init, Fore

# Wraps stdout and stderr in a proxy object that resets any terminal color after print()
init(autoreset=True)


class Log:
    """
    Simple console output utility
    """

    loglevel: int = DEBUG

    @staticmethod
    def debug(msg: str):
        """
        Print in a lightblue color
        :param str: The message
        """

        if Log.loglevel == DEBUG:
            print(f'{Fore.LIGHTBLUE_EX}{msg}')

    @staticmethod
    def server(event_type: str, msg: str):
        """
        Print a server message in a cyan color.
        :param event_type: The event type from the server payload
        :param msg: The message
        """

        print(f'{Fore.CYAN}[SERVER] ({event_type}) {msg}')

    @staticmethod
    def info(msg: str):
        """
        Print a regular unformatted message
        :param str: The message
        """

        if Log.loglevel <= INFO:
            print(msg)

    @staticmethod
    def warning(msg: str):
        """
        Print a warning message in yellow color.
        :param msg: The message
        """

        if Log.loglevel <= WARNING:
            print(f'{Fore.YELLOW}{msg}')

    @staticmethod
    def error(msg: str):
        """
        Print an error mesage to stderr in red
        :param msg: The message
        """

        if Log.loglevel <= ERROR:
            print(f'{Fore.RED}{msg}', file=stderr)

File: javascript/src/log.js
const chalk = require('chalk')


const DEBUG = 0
const INFO = 10
const WARNING = 20
const ERROR = 30


/**
 * Simple console output utility
 */
const Log = {
    logLevel: DEBUG,

    /**
     * Print in color
     * @param {*} msg The message 
     */
    debug(msg) {
        if (Log.logLevel === DEBUG) {
            console.log(chalk.cyan(msg))
        }
    },

    /**
     * Print a server message in a light blue color
     * @param {*} eventType The event type from the server payload
     * @param {*} msg The message
     */
    server(eventType, msg) {
        console.log(chalk.blueBright(`[SERVER] (${eventType}) ${msg}`))
    },

    /**
     * Print a regular unformatted message
     * @param {*} msg 
     */
    info(msg) {
        if (Log.logLevel <= INFO) {
            console.log(msg)
        }
    },

    /**
     * Print a warning message in yellow color
     * @param {*} msg The message
     */
    warning(msg) {
        if (Log.logLevel <= WARNING) {
            console.log(chalk.yellow(msg))
        }
    },

    /**
     * Print an error message in red color
     * @param {*} msg The message
     */
    error(msg) {
        if (Log.logLevel <= ERROR) {
            console.log(chalk.red(msg))
        }
    }
}


module.exports = {
    debug: Log.debug,
    info: Log.info,
    warning: Log.warning,
    error: Log.error,
    server: Log.server,
    DEBUG: DEBUG,
    INFO: INFO,
    WARNING: WARNING,
    ERROR: ERROR
}
